### PR TITLE
feat: Auto calculate `commentstring`

### PR DESCRIPTION
**Pull request by tjdevries**
Not all the way done, but figured I'd shoot this over to you so you can take a look and let me know what you think.

### CODE REVIEW

**Comment by numToStr** [comment]
Thanks a ton. I'll have a look at this on the weekend or so.

**Comment by tjdevries** [comment]
Should I be doing:

```lua
            -- TODO: Dunno if this is any good or not.
            return config[ctx.ctype] or config[1]
```

Or just return the `config[ctx.ctype]` and if it doesn't have block comment, then it's all done?

**Comment by numToStr** [comment]
> Or just return the `config[ctx.ctype]` and if it doesn't have a block comment, then it's all done?

I thought about this before and IMO we should throw an error if the block comment isn't found when doing `gb` and friends. Although this should be done in its separate PR.

---

One question, Is adding context awareness by default a good idea? I mean some (like me) don't use context-commentstring and for some language like rust, this is not necessarily useful to calculate comment string.

**Comment by tjdevries** [comment]
I can make it short circuit for languages that don't have additional configuration.

My thought would be is if you know this is definitely something people will use and you can make it in a way that doesn't require any additional configuration, why not just enable it? It might cost a few bonus cpu cycles, but you obviously can afford it. You compile rust code 😂

**Comment by numToStr** [comment]
> It might cost a few bonus cpu cycles, but you obviously can afford it

Yeah I guess we can do that 😄 

---

But enter `jsx`. And this is one of the most bizarre things in this world. This literally deserves its own plugin.

```jsx
// When only attribute is on the line, then `//%s` is used.
export function Element() {
  return (
    <div>
      <h1
        // id="foo"
        // class="lsp"
        data-telescope="good"
      >
        Hello, world!
      </h1>
    </div>
  );
}

// When attribute is on the same line as element, then `/*%s*/` is used.
export function Element() {
  return (
    <div>
      <h1 /* id="foo" class="lsp" */ data-telescope="good" >
        Hello, world!
      </h1>
    </div>
  );
}

// For element, then `{/*%s*/}` is used.
export function Element() {
  return (
    <div>
      {/*<h1 id="foo" class="lsp" data-telescope="good" >*/}
        Hello, world!
      </h1>
    </div>
  );
}
```
What do you think? Should we try to cover all these cases or let the `nvim-ts-context-commenstring` handle all this?

**Comment by sQVe** [comment]
I just wanted to leave my 0.02$ and voice that I think the marriage between `Comment.nvim` and ` nvim-ts-context-commentstring` works really well - especially now with the new changes in `type` and `location`.

**Comment by tjdevries** [comment]
What changes do you mean for `type` and `location`? I'm not familiar.

> What do you think? Should we try to cover all these cases or let the nvim-ts-context-commenstring handle all this?

I don't think you have to handle everything, but providing some framework to make things work reasonably well out of the box in what is essentially 70ish lines of code (the other bits are docs, config updates or renames effectively) seems like not a bad tradeoff. Of course, it's not my plugin :) so you can reject any time you want. I was just thinking this works surprisingly well with no dependencies (it only uses things from `vim.treesitter`, no deps otherwise).

Another thing is this calculates based on where your cursor currently is, so you can do interesting stuff like if you have multiple filetypes on one line, it will calculate the correct one, but you just get that for free basically from just checking the current node and walking up til you find either an interesting node or the root.

**Comment by tjdevries** [comment]
(Also, people can just do whatever crazy `jsx` config they want to make happen using `ft.set('jsx', { ... })`, and/or just using context comment string for that)

**Comment by numToStr** [comment]
> What changes do you mean for type and location? I'm not familiar.

He was talking about this https://github.com/JoosepAlviste/nvim-ts-context-commentstring/pull/28 change.

> I was just thinking this works surprisingly well with no dependencies (it only uses things from vim.treesitter, no deps otherwise)

The initial code that you showed on your stream, yeah that was also a total surprise to me as well. For something like vue or lua + ffi, this is all that you might need.

But IMO we should hold off adding this now. We can revisit this later :)

**Comment by arpangreat** [comment]
So when this pr is going to be merged ? I loving this plugin so far but this will be killer

**Comment by numToStr** [comment]
I am little caught up with my work right now. It could take a while before I merge this :)

**Comment by arpangreat** [comment]
> I am little caught up with my work right now. It could take a while before I merge this :)

Fine :) , BTW Great Work man ❤️

**Comment by tjdevries** [comment]
Anything I can do to help it get merged? I want to merge this before I make my TakeTuesday video about it :) Let me know if there is anything missing still for it.

**Comment by numToStr** [comment]
@tjdevries I am probably gonna merge it tomorrow morning :)

**Comment by tjdevries** [comment]
<3 excellent. Thanks. Maybe I will try and sneak in a recording session before this tuesday then :) thanks

File: lua/Comment/ts.lua [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    while contained and contained:type() and not config[contained:type()] do
+        contained = contained:parent()
+    end
+
+    -- Hmmm, might want this back, not sure yet.
+    -- if not contained then
+    --     return original
+    -- end
+
+    return contained
```
Comment (numToStr) [inline lua/Comment/ts.lua:106]: @tjdevries Is this supposed to be `return original`?

Comment (numToStr) [reply lua/Comment/ts.lua:106]: I've commented the line 96 and everything looks fine to me :)

**Comment by numToStr** [comment]
I've done the following changes
1. Added `vue` and `css` commentstring to please the vue devs.
2. Replaced `typescriptreact` with `tsx`. It seems treesitter returns `tsx` as a language for `typescriptreact` filetypes.
3. Removed `javascriptreact` commenstring. In this case treesitter returns `javascript` even for `javascriptreact` filetypes

File: lua/Comment/ts.lua [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+        range = {
+            cursor[1] - 1,
+            cursor[2],
+            cursor[1] - 1,
+            cursor[2],
+        }
+    end
+
+    return langtree:language_for_range(range), range
```
Comment (numToStr) [inline lua/Comment/ts.lua:53]: @tjdevries I think we have a range issue. You can see this in the following `vue` file.
```vue
<script>
export default {
  data() {
    return {
      greeting: "Hello World!",
    };
  },
};
</script>

<template>
  <p class="greeting">{{ greeting }}</p>
</template>

<style>
<!-- .greeting { -->     <---- This should be `css` but according to treesitter this is `vue`
  color: red;
  font-weight: bold;
}
/* </style> */           <---- This should be `vue` but according to treesitter this is `css`
```

Comment (tjdevries) [reply lua/Comment/ts.lua:53]: What kind of comment style? I could have an off-by-one still around somewhere that I hadn't considered

Comment (numToStr) [reply lua/Comment/ts.lua:53]: AFAIK, It should be 

```vue
<style>
/* .greeting { */
  color: red;
  font-weight: bold;
}
<!-- </style> -->
```

File: lua/Comment/ts.lua [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        elseif ctx.cmotion == U.cmotion.v then
+            local region = { U.get_region('v') }
+            cursor = { region[1] - 1, region[2] }
+        else
+            -- ctx.cmotion == U.cmotion.char
+            cursor = vim.api.nvim_win_get_cursor(0)
+        end
+
+        range = {
+            cursor[1] - 1,
```
[the remark this answers is not in the document]
Comment (numToStr) [reply lua/Comment/ts.lua:46]: Ahh, that seems to be the issue. I could try to fix this tomorrow morning If @tjdevries doesn't beat me to it :)

**Comment by tjdevries** [comment]
OK, I think I pushed. Checked a bunch of different combos in that file. Feel free to squash and merge or whatever you want to do w/ this after. Excited for it :) working on the video now too

**Comment by numToStr** [comment]
Thanks a ton ❤️. But, I am a little skeptical about this feature as I am pretty sure there will be some potential issue that we haven't found yet (some I did find). IMO 90% of the issues will always come from `jsx/tsx` trying to be cool. So, what we could do to support auto calculate? I am suggesting to only support proper/full-fledged languages that are injected/embedded something like HTML with `<script/>` tag for javascript and `<style/>` for CSS or Lua heredoc inside vimL. This, I guess, would cover most cases with injected languages.

But what about react/jsx gang? Ok, for that we could easily say that you should use [`nvim-ts-context-commentstring`](https://github.com/JoosepAlviste/nvim-ts-context-commentstring) as the builtin integration sucks for jsx. I would also like to discuss with the author to improve the integration story b/w mine and other plugins so that every user doesn't have to copy/paste 10-15 lines to get it working.

---

> only support proper/full-fledged languages that are injected/embedded

To do so I came with this (thanks to ts-playground)

```lua
function ft.calculate(ctype)
    local buf = vim.api.nvim_get_current_buf()
    local langtree = vim.treesitter.get_parser(buf)

    local win = vim.api.nvim_get_current_win()
    local row, col = unpack(vim.api.nvim_win_get_cursor(win))
    local range = { row - 1, col, row - 1, col }

    local found
    langtree:for_each_child(function(tree)
        if not found and tree:contains(range) then
            local lang = tree:lang()
            local cstr = ft.get(lang, ctype)
            if cstr then
                found = cstr
            end
        end
    end)

    return found or ft.get(vim.bo.filetype, ctype)
end
```
This is easy to maintain and reason about. (I am being selfish here 😅). And this is how this works with Lua (ffi), HTML and vue filetype.

https://user-images.githubusercontent.com/24727447/142155955-332d1ee9-f43d-448f-a5d0-d8a3e26e65b5.mp4

@tjdevries What's your opinion on this? I need your green light on this :)

**Comment by tjdevries** [comment]
Ah, so main difference is just that instead of being able to do for particular nodes, you'll only do it for embedded languages?

**Comment by tjdevries** [comment]
Hmm, I'm thinking about that part.

One thing I'm wondering, is are you sure you don't need this calculation?

```
    if not range then
        local cursor
        if ctx.cmotion == U.cmotion.line then
            cursor = get_cursor_line_non_whitespace_col_location()
        elseif ctx.cmotion == U.cmotion.V then
            local region = { U.get_region('V') }
            cursor = { region[1] - 1, region[2] }
        elseif ctx.cmotion == U.cmotion.v then
            local region = { U.get_region('v') }
            cursor = { region[1] - 1, region[2] }
        else
            -- ctx.cmotion == U.cmotion.char
            cursor = vim.api.nvim_win_get_cursor(0)
            cursor[1] = cursor[1] - 1
        end

        range = { cursor[1], cursor[2], cursor[1], cursor[2] }
    end
```

It seems that the "cursor" is not always in the right spot to actually calculate the language correctly.

**Comment by tjdevries** [comment]
(could easily be added to one function that is `get_currently_ts_range()` or something)

**Comment by numToStr** [comment]
From my little testing, I found no problem with it. IIRC that bit of code was only required for jsx but I think for embedded languages we don't need it.

**Comment by numToStr** [comment]
If you want I could update the PR. So everyone could test it. What do you say?

**Comment by tjdevries** [comment]
Yeah, go ahead and update the PR and we can test it out. I just was worried for some of the visual selection stuff.

The main reason I was asking was because if we do actually need a bit of complicated logic to calculate where the cursor is, the rest of my code is not super different from yours (right, it just checks name of node instead of just language).

But I can understand the hesitation about wanting to do too much for this, because users report bugs that they shouldn't report because you explicitly say "this don't work for jsx" haha

**Comment by numToStr** [comment]
Thanks! I'll update the PR.

> But I can understand the hesitation about wanting to do too much for this, because users report bugs that they shouldn't report because you explicitly say "this don't work for jsx"

Yup, that sums it up my feelings 😁. I'll rather prefer working on other treesitter stuff like smart uncommenting or something.

**Comment by tjdevries** [comment]
OK, works for me. Happy to say "keep using context comment string (for now) if you're using tsx".

Fortunately in my video I only showed it working for vue :laughing: I finished recording last night.

**Comment by numToStr** [comment]
> Fortunately in my video I only showed it working for vue 😆

If you showed jsx then it would've been a disaster 😂

**Comment by tjdevries** [comment]
haha I tested a ton of different combos on the last commit I had when I fixed the range and it all worked well for me, so disaster seems like the wrong word. :)

**Comment by numToStr** [comment]
Di you try this one. In vue file comment/uncomment everything **inside** a script tag.

**Comment by tjdevries** [comment]
like `gcit` ?

**Comment by numToStr** [comment]
Yes.

**Comment by tjdevries** [comment]
Don't remember -- I could test though if you want.

**Comment by numToStr** [comment]
Yup. FYI I also fixed the range issue locally today i then tested it `gcit` inside script but I wasn't able to uncomment. IDK maybe I didn't fixed it correctly 😂.

**Comment by tjdevries** [comment]
yes, uncomment won't work because treesitter doesn't work _within_ the comment, that will never work great really (as far as I can tell).

**Comment by tjdevries** [comment]
although, maybe it should have been able to do it from gcit... hmmm... interesting, idk

**Comment by numToStr** [comment]
Let me tell you, with my proposed code earlier it was working like coconut oil 😁. IDK how but it was working.

**Comment by tjdevries** [comment]
haha, ok :) let's merge yours and we can come back to this part of the issue later. I do not mind being unable to do it w/ node names (at least for now). Perhaps a different day I will think of a more elegant solution

**Comment by numToStr** [comment]
Sure, I'll work on this tomorrow (now bed time) 🙂

**Comment by seblj** [comment]
I tried it with gcit in vue, and it seems to be working most of the time IF you don't have a blank line inside the tag. Don't know why, but when removing the whitespace it seems to fix it. However it didn't fix it all the time. If you do `gcit` fast to first comment it and then immediately uncomment it, it uses `/* */` as commentstring. However, if you wait a second and then try to uncomment, it seems to be working consistently

**Comment by numToStr** [comment]
> Because when doing `vit` it seems like it adds a whitespace after the opening tag
@seblj This doesn't happen with me though.

BTW I am working on changing the logic for language detection. I think that will solve this issue. I'll let you know when it's done.

**Comment by numToStr** [comment]
@seblj @tjdevries I've added new treesitter detection. You can test it now. I am also working on improving the range detection i.e using the range of selection or text-objects (`gcit`) rather than using the current cursor position.

**Comment by seblj** [comment]
It still includes the opening tag. But I am pretty sure that this is just how it works, since I can recreate the behaviour with `vit` with a clean neovim with no plugins. (It's also this behaviour with vim). However, it fixes the problem where you comment, and them uncomment really fast like I mentioned in a previous comment

**Comment by numToStr** [comment]
> It still includes the opening tag. But I am pretty sure that this is just how it works, since I can recreate the behaviour with `vit` with a clean neovim with no plugins

Yup It does. If you have a whitespace after the opening tag.

**Comment by seblj** [comment]
> Yup It does. If you have a whitespace after the opening tag.

That's the thing, I don't have a whitespace. If you could try with this, and do a `vit` (or `gcit` to comment, you should see that it includes the starting tag even if you don't have a trailing whitespace) inside one of the tags that span over multiple lines. For example the style or script tags

```Vue
<template>
  <div>
    <p>Something</p>
  <div>
</template>

<script lang="ts">

import { defineComponent } from 'vue';

export default defineComponent({
  name: 'test',
    
  },
});
</script>

<style>
body {
  height: 100vh;
}
</style>
```

**Comment by seblj** [comment]
But this isn't a big deal since it is the same behaviour from other comment-plugins like for example `vim-commentary`. So this isn't something that should be "fixed" here IMO. If there is a fix needed, it should probably be to neovim

**Comment by numToStr** [comment]
@seblj I think something weird is going on with your nvim. Mine is working flawlessly.

https://user-images.githubusercontent.com/24727447/142403665-d2266d69-bb45-4884-be6d-4279a823e76f.mp4

(I have targets.vim installed, maybe that's why I am not getting that issue, IDK)

**Comment by seblj** [comment]
> EDIT: ~(I have targets.vim installed, maybe that's why I am not getting that issue, IDK)~ Yes, It was `targets.vim` that fixed the issue for me, I didn't know that 😄. I removed it and I got the same behavior as yours. And yeah, you are right this needs to be fixed in vim/neovim.

Oh okay, that's good to know! I might check out `targets.vim` later then. Glad we figured that out😄

**Comment by numToStr** [comment]
I have now refactored the range thingy. Apart from cleanup, I think everything is good to go. @seblj Care to test (and don't forget to install `targets.vim` 😆)?

PS: I might need to do some refactoring after this PR is merged.

**Comment by seblj** [comment]
> I have now refactored the range thingy. Apart from cleanup, I think everything is good to go. @seblj Care to test (and don't forget to install `targets.vim` 😆)?

There seems to be some issues when there are whitespaces in the tag-block. (I installed `targets.vim`).

https://user-images.githubusercontent.com/5160701/142409240-816df8eb-b863-4480-9042-d2d6fe64ef94.mov

**Comment by numToStr** [comment]
Hmm, I am getting this kinda similar issue in HTML file but vue is fine. I am looking into this.

**Comment by numToStr** [comment]
@seblj I get what's happening. Assuming the following style tag, doing `gcit` will not work.
```html
<style>
    .comment {
    ^ `css` starts from here
^ This whitespace area is `html`
        border: 1px solid red;
    }
</style>
```
> (could easily be added to one function that is get_currently_ts_range() or something)

Now I think @tjdevries was right with this one.

**Comment by numToStr** [comment]
> @seblj I get what's happening. Assuming the following style tag, doing `gcit` will not work.
> 
> ```
> <style>
>     .comment {
>     ^ `css` starts from here
> ^ This whitespace area is `html`
>         border: 1px solid red;
>     }
> </style>
> ```

Apart from this tag issue (which only happens where one language ends and the other starts), everything else looks good to me. We can still say you can use [`nvim-ts-context-commentstring`](https://github.com/JoosepAlviste/nvim-ts-context-commentstring) if you are not happy with the builtin one 😅.

**Comment by seblj** [comment]
Okay so I found out some of the issues and why they happen (I think). It seems to be the case like you say that treesitter is interpreting it as another language, but I could only reproduce this if the newline is between the starting tag and the first line inside the tags. 

```Vue
<script lang="ts">

import { defineComponent } from 'vue';
export default defineComponent({
  name: 'test',
  },
});
</script>
```

With the newline there, treesitter interprets that as filetype vue, and I don't think there is much we can do about that. One thing I think we can do is calculate the commentstring based on where the cursor is when doing the motion. Right now it looks like the cursor jumps to the start of the motion, and then when finding the language where the cursor is currently, the language is returned as vue. After this the cursor returns to where it was (probably has something to do with the sticky option). If one is holding the cursor over the newline and uncomment, there isn't much we can do (to my knowledge) since treesitter interprets this as vue.

However there is also a slight issue with empty lines if the commentstring surrounds the string (commentstrings like `/* %s */` for example). If there is a newline inside the tag when doing `gcit` for example, the newline will only get `/*` inserted on the line. This will mess with the logic of checking if a line is commented or not, so therefore it will try to comment this again instead of uncommenting.

```Vue
<style>
/* body { */
/*   height: 100vh; */
/*
/* } */
</style>
```

 Note that this will not be an issue with commentstrings like `// %s`, and also not for those who chose to ignore the newlines in lua pattern option in the setup function. I also don't think this is related to this PR, so this could probably be fixed in another PR.

If we add the rhs of the commentstring (`*/`) to the newline, after `/*` the problem is fixed. Like this

```Vue
<style>
/* body { */
/*   height: 100vh; */
/* */
/* } */
</style>
```

**Comment by numToStr** [comment]
> With the newline there, treesitter interprets that as filetype vue, and I don't think there is much we can do about that.

Yeah, that certainly a limitation. But as I said we can still rely on `nvim-ts-context-commenstring` to tackle this.

> However there is also a slight issue with empty lines if the commentstring surrounds the string (commentstrings like /* %s */ for example). If there is a newline inside the tag when doing gcit for example, the newline will only get /* inserted on the line. This will mess with the logic of checking if a line is commented or not, so therefore it will try to comment this again instead of uncommenting.

Ahh, yes I also noticed this sometimes. But this a general issue I can fix this later.

Anyways other than these I think we are good to go. I would like to merge this. What do you say?

**Comment by seblj** [comment]
> Yeah, that certainly a limitation. But as I said we can still rely on nvim-ts-context-commenstring to tackle this.

That is actually also a problem when using `nvim-ts-context-commentstring` so I think that issue maybe should be resolved by the treesitter parser of vue or injections in `nvim-treesitter`😅 (unless that is the intended behaviour🤷‍♂️)

> Anyways other than these I think we are good to go. I would like to merge this. What do you say?

Yes, other than those two, it looks to be working great

**Comment by numToStr** [comment]
Noice. I'll just cleanup some stuff and update readme then I'll merge this :)

**Comment by clason** [comment]
Just a heads-up: this PR completely broke comment.nvim for filetypes without a parser: you just get
```
E5108: Error executing lua ...local/share/nvim/runtime/lua/vim/treesitter/language.lua:25: no parser for 'tex'
 language, see :help treesitter-parsers
stack traceback:
        [C]: in function 'error'
        ...local/share/nvim/runtime/lua/vim/treesitter/language.lua:25: in function 'require_language'
        /usr/local/share/nvim/runtime/lua/vim/treesitter.lua:38: in function '_create_parser'
        /usr/local/share/nvim/runtime/lua/vim/treesitter.lua:93: in function 'get_parser'
        ...m/site/pack/packer/start/comment.nvim/lua/Comment/ft.lua:86: in function 'calculate'
        ...ite/pack/packer/start/comment.nvim/lua/Comment/utils.lua:217: in function 'parse_cstr'
        ...te/pack/packer/start/comment.nvim/lua/Comment/opfunc.lua:63: in function 'opfunc'
        ...e/pack/packer/start/comment.nvim/lua/Comment/comment.lua:166: in function <...e/pack/packer/start/c
omment.nvim/lua/Comment/comment.lua:165>
```

**Comment by numToStr** [comment]
Fixed 7627727

**Pull request merged**